provider/adguardhome/client.go

```diff
@@ -0,0 +1,226 @@
+		return nil, err
+	}
+
+	parsedURL, err := url.Parse(cfg.Server)
+	if err != nil {
+		return nil, err
+	}
+
+	httpClient := instrumented_http.NewClient(
+		&http.Client{
+			Transport: &http.Transport{
+				TLSClientConfig: &tls.Config{
+					InsecureSkipVerify: cfg.TLSInsecureSkipVerify,
+				},
+			},
+		},
+		&instrumented_http.Callbacks{})
+
+	client := &adGuardHomeClient{
+		BaseURL:    parsedURL,
+		cfg:        cfg,
+		httpClient: httpClient,
+	}
+
+	return client, nil
+}
+
+func (c *adGuardHomeClient) do(req *http.Request) (io.ReadCloser, error) {
+	req.Header.Add("content-type", "application/json")
+	req.SetBasicAuth(c.cfg.Username, c.cfg.Password)
+
+	res, err := c.httpClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	if res.StatusCode != http.StatusOK {
+		defer res.Body.Close()
+		return nil, fmt.Errorf("received non-200 status code from %s request: %s", res.Request.URL, res.Status)
+	}
+	return res.Body, nil
+}
+
+func (c *adGuardHomeClient) getType(entry *adGuardHomeEntry) string {
+	ip := net.ParseIP(entry.Answer)
+	if ip == nil {
+		return endpoint.RecordTypeCNAME
+	}
+	ip4 := ip.To4()
+	if ip4 != nil {
+		return endpoint.RecordTypeA
+	}
+
+	return endpoint.RecordTypeAAAA
+}
+
+func (c *adGuardHomeClient) listRecords(ctx context.Context) ([]*endpoint.Endpoint, error) {
+	url := c.BaseURL.ResolveReference(&url.URL{Path: "/control/rewrite/list"})
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url.String(), nil)
+	if err != nil {
+		return nil, err
+	}
+
+	body, err := c.do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer body.Close()
+	raw, err := io.ReadAll(body)
+	if err != nil {
+		return nil, err
+	}
+
+	var entries []adGuardHomeEntry
+	if err := json.Unmarshal(raw, &entries); err != nil {
+		return nil, fmt.Errorf("failed to parse JSON: %v", err)
+	}
+
+	out := make([]*endpoint.Endpoint, 0)
+	for _, rec := range entries {
+		if c.cfg.DomainFilter.IsConfigured() && !c.cfg.DomainFilter.Match(rec.Domain) {
+			log.Debugf("Skipping %s that does not match domain filter", rec.Domain)
+			continue
+		}
+		out = append(out, &endpoint.Endpoint{
+			DNSName:    rec.Domain,
+			Targets:    []string{rec.Answer},
+			RecordType: c.getType(&rec),
+		})
+	}
+	return out, nil
+}
+
+func (c *adGuardHomeClient) handleRecord(ctx context.Context, ep *endpoint.Endpoint, action string) error {
+	record := adGuardHomeEntry{
+		Domain: ep.DNSName,
+		Answer: ep.Targets[0],
+	}
+
+	url := c.BaseURL.ResolveReference(&url.URL{Path: "/control/rewrite/" + action})
+	body, err := json.Marshal(record)
+	if err != nil {
+		return err
+	}
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, url.String(), bytes.NewBuffer(body))
+	if err != nil {
+		return err
+	}
+
+	if _, err := c.do(req); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *adGuardHomeClient) createRecord(ctx context.Context, ep *endpoint.Endpoint) error {
+	return c.handleRecord(ctx, ep, "add")
```

Comment:
Should probably verify the record is of a supported RecordType. Otherwise TXT records might get created as CNAME or somesuch.

Reply:
Good catch! Fixed it :)